### ISSUE TITLE
Trigger NIA tasks based on Consul key/value (KV) changes

### ISSUE THREAD

**Issue by sandstrom** [hashicorp/consul-terraform-sync#150]
#### Description
It would be great if we could trigger tasks based on specific keys in the KV/store. Right now, it only seems to trigger based on changes to a service.

https://www.consul.io/docs/nia/installation/configuration#task
https://www.consul.io/docs/nia/installation/configuration

#### Use Cases
We have two groups of nodes, "green" and "blue". To avoid downtime we shift traffic from one to the other when doing deploys. This shift is handled by our load balancer (AWS ALB) using their rules.

We'd like to use a Consul KV pair `load-balancer/active-group` : `green` to trigger a Network Infrastructure Automation task that would update the load balancer rules, via Terraform, to shift traffic to green.

#### Alternative Solutions
While services have meta data, it seems that metadata is only "inherited" from service definitions on individual nodes. 

If we want to shift traffic from our blue group to our green group of nodes, we want one canonical place where that value is stored (group receiving traffic). It wouldn't make sense to make service definition changes on an individual node.

AFAIK we can't set service-level meta directly in the catalog, for that service globally.

**Comment by findkim** [comment]
@sandstrom thanks for taking the time submit this feature request. And I see 👍 are trickling in 🎉.

Yep you're correct in that tasks are triggered only on services right now, and agreed I believe that the catalog API is limited for updating the service-level metadata after it has been deployed.

Thanks for laying out the use case well. I think this feature request is valuable and am interested in more insight on how the KV pair would change. For your use case (or other ppl interested in this feature!) would the KV pair `load-balancer/active-group`:`green` be updated by a manual or automated process? What does that look like?

**Comment by sandstrom** [comment]
@findkim Thanks for thinking about this!

It could be both automated and manual. In our case probably manual initially, but could be an automated script kicked off by e.g. a new tagged release in Github in the future.

Consul's KV store is quite useful for deployment and infrastructure state, stuff that shouldn't go in your primary database. Things like:

- maintenance mode (true/false)
- currently deployed blue/green group (string value)
- rate limiting thresholds
- rate of requests/traffic sent to canary channel
- etc

Being able to use NIA for KV changes would unlock a lot of possibilities.

**Comment by findkim** [comment]
Awesome, thank for more info on your use case! This is a feature request we anticipated and have been collecting different use cases to help us design and prioritize a KV trigger.

**Comment by sandstrom** [comment]
@findkim Sounds great, let me know how it goes!

**Comment by findkim** [comment]
> Things like these would be great to control via NIA tasks:
> 
> * maintenance mode (true/false)
> * currently deployed blue/green group (string value)
> * rate limiting thresholds
> * rate of requests/traffic sent to canary channel (e.g. 10% of traffic to canary channel)

I'm trying to look at the use-cases at a high level, and there are 2 common themes I'm seeing:
* An interactive feature to update values that a task relies on
* Another means to trigger task execution aside from service updates

Those example values listed seem like good candidates for module input variables. I'm wondering if an interactive way to update input variables could be an alternative to KV triggers. A very rough example to show what I'm thinking, a CLI or REST API that be used to interact with a running CTS.

```
consul-terraform-sync task update <task-name> -var "active_group=green"
```

where `active_group` is an input variable for the module and the command would trigger the task. These in-memory updates would be ephemeral based on the running CTS process. On restart it would default to the module's default value or the value set in `task.variable_files`. With this route, the UX remains within CTS and is the central location for these options.

@sandstrom you had mentioned you want "one canonical place" -- are there other applications that need to know about it, for example whether blue or green is the active group aside from the updated infrastructure?

**Comment by sandstrom** [comment]
I think we'd actually prefer to not interact with the running CTS directly. The CTS nodes will probably have elevated permissions, since it's gotta be able to trigger terraform changes. 

It would be good if that node was fairly self-sustaining, i.e. didn't need people or processes to sign into it, and do stuff. Better if it only acts on changes in consul.

Also, the benefit with using Consul KV to store e.g. active group is that other systems can query it + we can use existing ACL in Consul to restrict what entities can update certain KV keys.

To your specific suggestion about `consul-terraform-sync task update <task-name> -var "active_group=green"`. Sorry to be blunt, but what's the point of doing that? 

In that case we can just as well do a vanilla `terraform apply -var="active_group=green"`. I don't see what value CTS would bring in that case.

**Comment by findkim** [comment]
Of course all of CTS can be done with vanilla Terraform, that's no secret. The software layer of CTS that offers discovery of changes at scale for automation is the value.

As for the command, I'm not set on that UX exactly, but I don't see it much different than the extra step of updating the KV. An asynchronous step is needed for a value to change to trigger execution. But with KV it requires the operator(s) to know the key and value format in 3 different places: CTS, modules, and now KV. Once it's set up to work it'll work well. But an invalid value or wrong key would be very unfriendly to debug -- update KV and hope for CTS downstream picks it up and is the right format for the module. Working directly with CTS you would be able to receive feedback directly -- validity, errors, success, etc. Adding a layer of security/permissions around CTS is not out of the question either. An additional option could have CTS write those values also upstream to KV for visibility of other systems, keeping CTS the central control of NIA.


In any case,

> prefer to not interact with running CTS

I find this an interesting perspective and can understand it from elevated permissions standpoint, but a means to execute a task outside of Consul changes is likely a scoped topic for a different feature. So thanks for your input on how it would not address your KV request. This ticket can continue on the topic of KV triggers

**Comment by mkam** [comment]
Completed this feature which will be released as part of v0.4.0.

**Comment by sandstrom** [comment]
@mkam Awesome! 🎉 

Couldn't find a lot of documentation, but maybe that's coming in a future PR? (or I'm just not good at looking)

**Comment by mkam** [comment]
@sandstrom The documentation isn't live yet since we haven't released v0.4.0, but if you want a preview of the related docs, you can check this pull request out: https://github.com/hashicorp/consul/pull/11025

**Comment by sandstrom** [comment]
thanks @mkam

**Comment by mkam** [comment]
@sandstrom This has been released as part of our beta release (v0.4.0-beta1) if you want to try it out! The documentation is available on consul.io now as well.
